### PR TITLE
Sync XMLUtils with kodi master

### PR DESCRIPTION
**Pull request by manuelm**
This replaces the usage of `CStdString` with `std::string` so binary addons can get rid of `CStdString` while still using `XMLUtils` and socket abstraction.

I've synced `XMLUtils` with Kodi master. This adds a few new methods but also removes two xml-encoding helper methods. These methods are not used in any binary addon so I've removed them as well.
Additional the sync fixes a small bug in `XMLUtils::GetString` returning false if node-value is empty (e.g. `<foo/>`)

On linux all binary addons build fine with pvr.wmc being the exception. Adding `#include "kodi/util/StdString.h"` to pvr.wmc/src/utilities.cpp will fix it. I'll create a PR for pvr.wmc if this PR has been accepted.

### CODE REVIEW

**Comment by opdenkamp** [comment]
libCEC is still using the original code, and the platform lib should be split up into a Kodi-specific part and the actual libplatform first, or I suspect that changing signatures like this will cause issues when libCEC and add-ons are loaded by Kodi, and both have a different platform lib.

**Comment by manuelm** [comment]
yeah, possible. But we can postpone the socket changes.

What about the XMLUtils changes/sync? libCEC isn't using them.

**Comment by opdenkamp** [comment]
xmlutils is fine indeed, that's Kodi-specific

**Comment by manuelm** [comment]
The socket commit is gone.

Will create a PR to update kodi-platform revision after this is merged.

**Comment by Montellese** [comment]
Looks good to me.

**Comment by manuelm** [comment]
@opdenkamp Any further objections?

**Comment by opdenkamp** [comment]
no, it's fine by me

FYI @Montellese: I've stripped all the Kodi and libCEC specific stuff from this library, and renamed `kodi-platform` to `platform`: https://github.com/Pulse-Eight/platform
PLATFORM namespace is unchanged.
this will be used by the next update of libCEC after this cmake PR has been merged in: https://github.com/Pulse-Eight/libcec/pull/65
`kodi-platform` should be the combination of this one with xml-utils, then nothing has to change in add-ons that are using it.

**Comment by wsnipex** [comment]
@opdenkamp libplatform is not really a good name.
apt-cache search libplatform

**Comment by opdenkamp** [comment]
I did that, and got some `libplatform-api` and related files, not `libplatform`

**Comment by opdenkamp** [comment]
the libCEC PR that I linked needs some work on Windows btw after dropping `CStdString` and currently won't compile on that os.

**Comment by wsnipex** [comment]
yes, libplatform-api and libplatform still seem very similar and will cause at least confusion.
Can't you name it less ambiguous?

**Comment by opdenkamp** [comment]
well, it's always been called `libplatform` cause it used to sit in `src\lib\platform` ;)

this `platform-api` is some ubuntu-only package, part of some preview for phablets (whatever that may be), and not used by anything but ubuntu from the looks of it

**Comment by manuelm** [comment]
Can someone with sufficient permissions please merge this?

**Comment by FernetMenta** [comment]
this is not a fix and we are not in a merge window

**Comment by manuelm** [comment]
@FernetMenta Can you please point me to a roadmap or a timeline?

My plan was to wait for this PR before creating a PR for my PVR. I've about 15 commits ready. Some are fixes, some are new features or just code refactoring. What about those?

**Comment by FernetMenta** [comment]
the merge window is from 1st to 20th during alpha. only fixes are allowed outside this window.

**Comment by hudokkow** [comment]
What's the status on this? It would be nice to get rid of CStdString for 16

**Comment by opdenkamp** [comment]
Status = I'm on 'holiday' for another week :)

-------- Oorspronkelijk bericht --------
Van: "h.udo" notifications@github.com 
Datum:27-07-2015  18:28  (GMT+01:00) 
Aan: xbmc/kodi-platform kodi-platform@noreply.github.com 
Cc: Lars Op den Kamp opdenkamp@gmail.com 
Onderwerp: Re: [kodi-platform] Replace usage of CStdString with std::string (#2) 

What's the status on this? It would be nice to get rid of CStdString for 16

**Comment by hudokkow** [comment]
Yeah, I know. Happy renovation! :P

@manuelm can you rebase?

**Comment by manuelm** [comment]
Rebased + synced with latest XMLUtils from kodi master.

**Comment by hudokkow** [comment]
@manuelm, thanks!

**Comment by wsnipex** [comment]
ping, guess this has been overlooked?

File: src/util/XMLUtils.cpp [hunk 1]
```diff
@@ -13,20 +13,22 @@
...
- *  the Free Software Foundation, 675 Mass Ave, Cambridge, MA 02139, USA.
- *  http://www.gnu.org/copyleft/gpl.html
+ *  along with XBMC; see the file COPYING.  If not, see
+ *  <http://www.gnu.org/licenses/>.
  *
  */
 
 #include "XMLUtils.h"
+#include "StringUtils.h"
+#ifdef TARGET_WINDOWS
```
Comment (opdenkamp) [inline src/util/XMLUtils.cpp:23]: looks like this one is sitting in kodi's tree and needs to be moved over here

File: src/util/XMLUtils.cpp [hunk 2]
```diff
@@ -13,20 +13,22 @@
...
+#include "PlatformDefs.h" //for strcasecmp
+#endif
 
 bool XMLUtils::GetHex(const TiXmlNode* pRootNode, const char* strTag, uint32_t& hexValue)
 {
   const TiXmlNode* pNode = pRootNode->FirstChild(strTag );
   if (!pNode || !pNode->FirstChild()) return false;
-  sscanf(pNode->FirstChild()->Value(), "%x", (uint32_t*) &hexValue );
-  return true;
+  return sscanf(pNode->FirstChild()->Value(), "%x", (uint32_t*)&hexValue) == 1;
```
Comment (opdenkamp) [inline src/util/XMLUtils.cpp:31]: this should be in a separate commit. and this cast is nasty.

File: src/util/XMLUtils.cpp [hunk 3]
```diff
@@ -38,6 +40,17 @@ bool XMLUtils::GetUInt(const TiXmlNode* pRootNode, const char* strTag, uint32_t&
   return true;
 }
 
+bool XMLUtils::GetUInt(const TiXmlNode* pRootNode, const char* strTag, uint32_t &value, const uint32_t min, const uint32_t max)
```
Comment (opdenkamp) [inline src/util/XMLUtils.cpp:43]: should be in a separate commit too, this isn't replacing CStdString

File: src/util/XMLUtils.cpp [hunk 4]
```diff
@@ -109,18 +122,25 @@ bool XMLUtils::GetBoolean(const TiXmlNode* pRootNode, const char* strTag, bool&
...
+    strStringValue = pNode->ValueStr();
     return true;
   }
-  strStringValue.Empty();
-  return false;
+  strStringValue.clear();
+  return true;
+}
+
+std::string XMLUtils::GetString(const TiXmlNode* pRootNode, const char* strTag)
```
Comment (opdenkamp) [inline src/util/XMLUtils.cpp:139]: more new stuff. is this a copy + paste from somewhere else?

File: src/util/XMLUtils.cpp (hunk 2 above)
Comment (Montellese) [inline src/util/XMLUtils.cpp:31]: The cast comes from the code in kodi code. We should keep those in sync.
I agree that the sync of XMLUtils should be in its own commit.

**Comment by opdenkamp** [comment]
this is a copy from https://github.com/xbmc/xbmc/blob/master/xbmc/utils/XMLUtils.cpp
it must be moved over here instead with it's deps and have Kodi use that instead.
the whole point of these libraries is to get rid of the multiple copies of the same file that are all a little different, which is what we had before.

File: src/util/XMLUtils.cpp (hunk 4 above)
Comment (manuelm) [inline src/util/XMLUtils.cpp:139]: The name of the PR is misleading. The commit msg is `Sync XMLUtils with kodi master` and that's what I've done. As far as I remember I haven't added a single line myself.

**Comment by Montellese** [comment]
It's not exactly the same. The dependency on `CURL` is missing and therefore also the possibility of some of the methods to handle encoded URLs.

File: src/util/XMLUtils.cpp (hunk 4 above)
Comment (opdenkamp) [inline src/util/XMLUtils.cpp:139]: Ah yeah I was looking at the title of this window :)

**Comment by opdenkamp** [comment]
Isn't it a better idea to split things up properly then instead of trying to keep the files in sync? Or add the CURL dep to kodi-platform and move it over completely?

File: src/util/XMLUtils.cpp (hunk 2 above)
Comment (opdenkamp) [inline src/util/XMLUtils.cpp:31]: Then this cast in Kodi isn't very nice either ;-)

**Comment by Montellese** [comment]
I agree that things should be better and properly split up and self-contained but that's a very long and hard way to go :-/ `CURL` even depends on `CApplication`, `CLog` and some VFS stuff. All "design flaws" and hacks.
Personally I think it's not the concern of a utils class like `XMLUtils` to be able to URL decode anything. I also don't know if we use it anywhere in Kodi.

**Comment by opdenkamp** [comment]
guess we better wait with this until krypton+1
